libvips/include/vips/stream.h

```diff
@@ -0,0 +1,247 @@
+
+#ifdef __cplusplus
+extern "C" {
+#endif /*__cplusplus*/
+
+#define VIPS_TYPE_STREAM (vips_stream_get_type())
+#define VIPS_STREAM( obj ) \
+	(G_TYPE_CHECK_INSTANCE_CAST( (obj), \
+	VIPS_TYPE_STREAM, VipsStream ))
+#define VIPS_STREAM_CLASS( klass ) \
+	(G_TYPE_CHECK_CLASS_CAST( (klass), \
+	VIPS_TYPE_STREAM, VipsStreamClass))
+#define VIPS_IS_STREAM( obj ) \
+	(G_TYPE_CHECK_INSTANCE_TYPE( (obj), VIPS_TYPE_STREAM ))
+#define VIPS_IS_STREAM_CLASS( klass ) \
+	(G_TYPE_CHECK_CLASS_TYPE( (klass), VIPS_TYPE_STREAM ))
+#define VIPS_STREAM_GET_CLASS( obj ) \
+	(G_TYPE_INSTANCE_GET_CLASS( (obj), \
+	VIPS_TYPE_STREAM, VipsStreamClass ))
+
+/* Communicate with something like a socket or pipe. 
+ */
+typedef struct _VipsStream {
+	VipsObject parent_object;
+
+	/*< private >*/
+
+	/* Read/write this fd if connected to a system pipe/socket. Override
+	 * ::read() and ::write() to do something else.
+	 */
+	int descriptor;	
+
+	/* A descriptor we close with vips_tracked_close().
+	 */
+	int tracked_descriptor;	
+
+	/* A descriptor we close with close().
+	 */
+	int close_descriptor;	
+
+	/* If descriptor is a file, the filename we opened. Handy for error
+	 * messages. 
+	 */
+	char *filename; 
+
+} VipsStream;
+
+typedef struct _VipsStreamClass {
+	VipsObjectClass parent_class;
+
+} VipsStreamClass;
+
+GType vips_stream_get_type( void );
+
+const char *vips_stream_filename( VipsStream *stream );
+
+#define VIPS_TYPE_STREAM_INPUT (vips_stream_input_get_type())
+#define VIPS_STREAM_INPUT( obj ) \
+	(G_TYPE_CHECK_INSTANCE_CAST( (obj), \
+	VIPS_TYPE_STREAM_INPUT, VipsStreamInput ))
+#define VIPS_STREAM_INPUT_CLASS( klass ) \
+	(G_TYPE_CHECK_CLASS_CAST( (klass), \
+	VIPS_TYPE_STREAM_INPUT, VipsStreamInputClass))
+#define VIPS_IS_STREAM_INPUT( obj ) \
+	(G_TYPE_CHECK_INSTANCE_TYPE( (obj), VIPS_TYPE_STREAM_INPUT ))
+#define VIPS_IS_STREAM_INPUT_CLASS( klass ) \
+	(G_TYPE_CHECK_CLASS_TYPE( (klass), VIPS_TYPE_STREAM_INPUT ))
+#define VIPS_STREAM_INPUT_GET_CLASS( obj ) \
+	(G_TYPE_INSTANCE_GET_CLASS( (obj), \
+	VIPS_TYPE_STREAM_INPUT, VipsStreamInputClass ))
+
+/* Read from something like a socket, file or memory area and present the data
+ * with a simple seek / read interface.
+ *
+ * During the header phase, we save data from unseekable sources in a buffer
+ * so readers can rewind and read again. We don't buffer data during the
+ * decode stage.
+ */
+typedef struct _VipsStreamInput {
+	VipsStream parent_object;
+
+	/* We have two phases: 
+	 *
+	 * During the header phase, we save bytes read from the input (if this
+	 * is an unseekable source) so that we can rewind and try again, if
+	 * necessary.
+	 *
+	 * Once we reach decode phase, we no longer support rewind and the
+	 * buffer of saved data is discarded.
+	 */
+	gboolean decode;
+
+	/* TRUE is this descriptor supports lseek(). If not, then we save data
+	 * read during header phase in a buffer.
+	 */
+	gboolean seekable;
+
+	/*< private >*/
+
+	/* The current read point.
+	 */
+	off_t read_position;
```

Comment:
Perhaps this should be `gint64`? This ensures that we can handle large streams on 32-bit machines.

Reply:
Good idea!

Reply:
... do you know if lseek() returns a 32-bit value on 32-bit Windows? We might need to make an lseek() wrapper that calls whatever their 64-bit API is, just as we had to for vips_file_length().

Reply:
I think `lseeki64` is only able to do 64-bit seeks. I noticed there is a `vips__seek` function:
https://github.com/libvips/libvips/blob/be7c1404c321917c602c1fb3f54da009a92bef0d/libvips/iofuncs/util.c#L1031

Perhaps we could use that?

Reply:
Oh duh, I totally forgot about that. Yes, I'll fix it up and use it. Thanks!